scene/gui/tree.cpp

```diff
@@ -3912,6 +3997,12 @@ bool Tree::get_allow_reselect() const {
 	return allow_reselect;
 }
 
+void Tree::set_collapsed_all(bool p_collapsed, bool p_ignore_active) {
+
+	get_root()->set_collapsed_recursive(p_collapsed, p_ignore_active, false);
+	ensure_cursor_is_visible();
```

Comment:
Why needs the cursor to be always visible? I could call that method programmatically where I wouldn't want the cursor to be visible

Reply:
The name is confusing, but this method was already there and is not really about the cursor itself, but rather what's underneath the cursor. It's actually about the selected TreeItem.

If I don't call "ensure_cursor_is_visible()" when unfolding, the TreeItem that was selected by the user might end up out of view. If the tree is big and requires scrolling both horizontally and vertically, it could end up scrolling the view completely to the right and down or anywhere else depending on the last operations the user did before collapsing everything. 

You normally want the item the user interacted with last to stay visible. 

I usually would agree though, I don't like enforcing things like that, but the goal of this method is to make the "Collapse All" logic the same everywhere across the editor so the UX stays consistent. 

The main use case for this method would be: 
RightClick a TreeItem, which selects it and open a context menu -> select in the menu "Collapse All" or "Extend All". 

Here's a quick example without ensure_cursor_is_visible():
![Ensure Cursor Visible False](https://user-images.githubusercontent.com/52422077/82078846-d37f4d00-96af-11ea-9aed-f39eacb932ca.gif)

And here's with ensure_cursor_is_visible():
![Ensure Cursor Visible True](https://user-images.githubusercontent.com/52422077/82079746-5a80f500-96b1-11ea-9c4c-0c482dcc844b.gif)

I'm open to changing it if needed, but if we do, maybe we should instead add a third boolean property like "skip_ensure_cursor_is_visible" so devs automatically know it's **supposed** to be the default behaviour?

What do you think?

Reply:
So if I understood correctly, the method is all about ensuring the selected item is still in view. (Well as stated in the class reference...). I wouldn't add a third parameter to disable the behaviour, since I can't imagine a case, where it is desired to end up scrolled anywhere.

Reply:
FYI https://github.com/godotengine/godot/issues/16863#issuecomment-629639167